hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/commit/staging/StagingCommitter.java

```diff
@@ -0,0 +1,908 @@
+import static org.apache.hadoop.fs.s3a.commit.staging.StagingCommitterConstants.*;
+import static org.apache.hadoop.fs.s3a.commit.CommitConstants.*;
+import static org.apache.hadoop.fs.s3a.commit.CommitUtils.*;
+
+/**
+ * Committer based on the contributed work of the
+ * <a href="https://github.com/rdblue/s3committer">Netflix multipart committers.</a>
+ * <ol>
+ *   <li>
+ *   The working directory of each task is actually under a temporary
+ *   path in the local filesystem; jobs write directly into it.
+ *   </li>
+ *   <li>
+ *     Task Commit: list all files under the task working dir, upload
+ *     each of them but do not commit the final operation.
+ *     Persist the information for each pending commit into the cluster
+ *     for enumeration and commit by the job committer.
+ *   </li>
+ *   <li>Task Abort: recursive delete of task working dir.</li>
+ *   <li>Job Commit: list all pending PUTs to commit; commit them.</li>
+ *   <li>
+ *     Job Abort: list all pending PUTs to commit; abort them.
+ *     Delete all task attempt directories.
+ *   </li>
+ * </ol>
+ */
+public class StagingCommitter extends AbstractS3GuardCommitter {
+
+  private static final Logger LOG = LoggerFactory.getLogger(
+      StagingCommitter.class);
+  public static final String NAME = "StagingCommitter";
+  private final Path constructorOutputPath;
+  private final long uploadPartSize;
+  private final String uuid;
+  private final boolean uniqueFilenames;
+  private final FileOutputCommitter wrappedCommitter;
+
+  private ConflictResolution conflictResolution;
+  private final Path finalOutputPath;
+  private String s3KeyPrefix = null;
+
+  /** The directory in the cluster FS for commits to go to. */
+  private Path commitsDirectory;
+
+  /**
+   * Committer for a single task attempt.
+   * @param outputPath final output path
+   * @param context task context
+   * @throws IOException on a failure
+   */
+  public StagingCommitter(Path outputPath,
+      TaskAttemptContext context) throws IOException {
+    super(outputPath, context);
+    this.constructorOutputPath = checkNotNull(getOutputPath(), "output path");
+    Configuration conf = getConf();
+    this.uploadPartSize = conf.getLongBytes(
+        MULTIPART_SIZE, DEFAULT_MULTIPART_SIZE);
+
+    // Spark will use a fake app ID based on the current minute and job ID 0.
+    // To avoid collisions, use the YARN application ID for Spark.
+    this.uuid = getUploadUUID(conf, context.getJobID());
+    this.uniqueFilenames = conf.getBoolean(
+        FS_S3A_COMMITTER_STAGING_UNIQUE_FILENAMES,
+        DEFAULT_STAGING_COMMITTER_UNIQUE_FILENAMES);
+    setWorkPath(buildWorkPath(context, uuid));
+    this.wrappedCommitter = createWrappedCommitter(context, conf);
+    // forces evaluation and caching of the resolution mode.
+    ConflictResolution mode = getConflictResolutionMode(getJobContext());
+    LOG.debug("Conflict resolution mode: {}", mode);
+    this.finalOutputPath = constructorOutputPath;
```

Comment:
Not being able to change the output directory in a subclass makes it so we can't use this committer as the base for our batch committers. That's not a big deal for us, since we will probably continue using our own for a while. But, it would be nice to be able to change the output location based on table metadata. Our directory committer, for example, will create a new directory for data instead of deleting the current path, then it will switch the table location to the new directory. That can't be done unless there is a way to hook this output path.

Reply:
Ok, I'll make it non-final :)

Reply:
Update: next revision will 
1. eliminate `finalOutputPath`as a field
1. call `setOutputPath(finalOutputPath)` in constructor
1. then call `getOutputPath()` and use it in the constructor to set up the destFS, which is always fetched with `getDestFS()`
1. And all uses of `getOutputPath(JobContext)` are gone (in job setup and commit); expect the committer to be set up with the correct output path.

This should mean that whatever sets output path will define that output path,; it should be overrideable, provided that the constructor-invoked methods don't cause problems before a subclass is inited. Will that work?